src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -202,20 +207,42 @@ private void updateOriginalQueryResults(final CombineScoresDto combineScoresDTO,
         final List<CompoundTopDocs> queryTopDocs = getCompoundTopDocs(combineScoresDTO, querySearchResults);
         final Sort sort = combineScoresDTO.getSort();
         int totalScoreDocsCount = 0;
-        for (int index = 0; index < querySearchResults.size(); index++) {
-            QuerySearchResult querySearchResult = querySearchResults.get(index);
-            CompoundTopDocs updatedTopDocs = queryTopDocs.get(index);
-            totalScoreDocsCount += updatedTopDocs.getScoreDocs().size();
-            TopDocsAndMaxScore updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(
-                buildTopDocs(updatedTopDocs, sort),
-                maxScoreForShard(updatedTopDocs, sort != null)
+
+        // Get index of first non-empty CompoundTopDocs to check if collapse is enabled
+        boolean isCollapseEnabled = false;
+        int indexOfFirstNonEmpty = 0;
+        for (CompoundTopDocs compoundTopDocs : queryTopDocs) {
+            if (!compoundTopDocs.getTopDocs().isEmpty() && compoundTopDocs.getTopDocs().getFirst() instanceof CollapseTopFieldDocs) {
+                isCollapseEnabled = true;
+                break;
+            }
+            indexOfFirstNonEmpty++;
+        }
+        if (isCollapseEnabled) {
+            totalScoreDocsCount = collapse(
```

Comment:
Here we need totalScoreDocsCount But the method which returns it is named as collapse. Can you follow a principle of every method should do one specific type of operation and should be returning the information related to it?

Single Responsibility principle.